### ISSUE TITLE
Change provenance checking from stop on error, to simply write a warning...

### ISSUE THREAD

**Issue by ekluzek** [ESMCI/cime#2430]
The call to manage_externals to establish status for provenance information requires success. I think this should allow it to fail, and for it to report it had trouble finding the info.

@rgknox did have trouble with this. We have asked him to replicate the problem.

Here's the current code in CIME/provenance.py
```
def _save_build_provenance_cesm(case, lid): # pylint: disable=unused-argument
    version = case.get_value("MODEL_VERSION")
    # version has already been recorded
    srcroot = case.get_value("SRCROOT")
    manic = os.path.join(srcroot, "manage_externals","checkout_externals")
    out = None
    if os.path.exists(manic):
        out = run_cmd_no_fail(manic + " --status --verbose", from_dir=srcroot)
```

replacing run_cmd_no_fail with "run_cmd" would have it NOT die on an error, also give some information about the error it ran into.

Leaving it as it is, means we require it to work no matter what. Changing it does mean that sometimes the provenance information wouldn't be correct if something went wrong. So there is more than one way to look at this. So I'd like some discussion and then we can set it the way that we think makes the most sense.

**Comment by jedwards4b** [comment]
Can you open an issue in manage_externals and document what caused it to fail.  I've been trying to make sure that this doesn't happen.   Also make sure that this is actually broken at the head of the manage_externals master - some problems have already been addressed.

**Comment by ekluzek** [comment]
Here's the error that @rgknox got below. From looking at it, it's because his Externals file was screwed up. I think he had a branch name for the tag keyword. So again the question is do we want it to fail when there is some problem like this in finding provenance information because a manage_externals file is screwed up. Or do we want it to go on and just report a problem?
```
Finished MODEL_BUILD for test ERS_D_Ld5.f19_g16.I2000Clm50BgcCruGs.lawrencium-lr3_intel.clm-default in 207.398784 seconds (FAIL). [COMPLETED 1 of 16]
    Case dir: /global/scratch/rgknox/ctsm-tests/fates.lawrencium-lr3.intel.Cd484ca2-F5c40b26-ngeet-master/ERS_D_Ld5.f19_g16.I2000Clm50BgcCruGs.lawrencium-lr3_intel.clm-default.G.20180315_150625_084n23
    Errors were:
        Building test for ERS in directory /global/scratch/rgknox/ctsm-tests/fates.lawrencium-lr3.intel.Cd484ca2-F5c40b26-ngeet-master/ERS_D_Ld5.f19_g16.I2000Clm50BgcCruGs.lawrencium-lr3_intel.clm-default.G.20180315_150625_084n23
        Exception during build:
        ERROR: Command: '/global/scratch/rgknox/ctsm/manage_externals/checkout_externals --status --verbose' failed with error 'Processing externals description file : Externals.cfg
        Processing externals description file : Externals_CISM.cfg
        Processing externals description file : Externals_CLM.cfg
        Checking status of externals: cism, source_cism, clm, fates, 
        ERROR: DEV_ERROR: regex to detect tracking branch failed.' from dir '/global/scratch/rgknox/ctsm'
        Traceback (most recent call last):
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/SystemTests/system_tests_common.py", line 88, in build
            model_only=(phase_name==MODEL_BUILD_PHASE))
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/SystemTests/system_tests_common.py", line 119, in build_phase
            self.build_indv(sharedlib_only=sharedlib_only, model_only=model_only)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/SystemTests/system_tests_common.py", line 126, in build_indv
            sharedlib_only=sharedlib_only, model_only=model_only)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/build.py", line 584, in case_build
            return run_and_log_case_status(functor, "case.build", caseroot=caseroot)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/utils.py", line 1496, in run_and_log_case_status
            rv = func()
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/build.py", line 583, in <lambda>
            functor = lambda: _case_build_impl(caseroot, case, sharedlib_only, model_only, buildlist)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/build.py", line 533, in _case_build_impl
            post_build(case, logs, build_complete=not (buildlist or sharedlib_only))
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/build.py", line 578, in post_build
            save_build_provenance(case, lid=lid)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/provenance.py", line 102, in save_build_provenance
            _save_build_provenance_cesm(case, lid)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/provenance.py", line 86, in _save_build_provenance_cesm
            out = run_cmd_no_fail(manic + " --status --verbose", from_dir=srcroot)
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/utils.py", line 424, in run_cmd_no_fail
            expect(False, "Command: '{}' failed with error '{}' from dir '{}'".format(cmd, errput.encode('utf-8'), os.getcwd() if from_dir is None else from_dir))
          File "/global/scratch/rgknox/ctsm/cime/scripts/Tools/../../scripts/lib/CIME/utils.py", line 112, in expect
            raise exc_type(msg)
        SystemExit: ERROR: Command: '/global/scratch/rgknox/ctsm/manage_externals/checkout_externals --status --verbose' failed with error 'Processing externals description file : Externals.cfg
        Processing externals description file : Externals_CISM.cfg
        Processing externals description file : Externals_CLM.cfg
        Checking status of externals: cism, source_cism, clm, fates, 
        ERROR: DEV_ERROR: regex to detect tracking branch failed.' from dir '/global/scratch/rgknox/ctsm'
```

**Comment by billsacks** [comment]
It feels like there are two questions:

(1) (Should be moved to manage_externals issue): Under what circumstances, if any, should `checkout_externals --status` give an error?

(2) What should the provenance code do when `checkout_externals` gives an error?

I feel like it's going to be hard / undesirable to completely eliminate the possibility of errors from `checkout_externals --status`. For example, what if there is a totally garbage `Externals.cfg` file? So (2) will be relevant in some cases. I personally don't have strong feelings as to whether there should be a warning or an error in these cases – it all depends on how important we feel it is that provenance information be captured: if it's important, then this should result in an error.

**Comment by billsacks** [comment]
(@ekluzek and @rgknox I know my current feelings don't totally match what I said on Thursday, when I think I supported turning this into a warning. Upon further reflection I'm thinking that maybe an error really is appropriate here - but again, I don't have strong feelings about this.)

**Comment by ekluzek** [comment]
I can go either way on this. I do think for important simulations I would like the provenance to be there for the case. So there are times I would want it to error out. But, it's also not obvious that you can't run a case or a test because the externals file for manage_externals are screwed up in some manner. And for a test or most debugging simulations I don't care if the provenance information isn't set.

I created an issue for this in manage_externals..

https://github.com/NCAR/manage_externals/issues/89

**Comment by billsacks** [comment]
@ekluzek I feel it's important for `manage_externals` to return an error code in some cases, like when the externals file is really messed up - i.e., saying, "I was not able to successfully determine the status". Then it's up to the caller to determine how to handle this. The provenance function could decide to proceed despite the error, or it could be changed to give a more meaningful error message if you feel that would help resolve this problem.

**Comment by jedwards4b** [comment]
Wait - this is a legitimate error then - I think that manage_externals and create_newcase did the right thing here.

**Comment by rgknox** [comment]
Hi All, 

This was the original problem I "thought" I had, and the steps I "thought" I took:
1) cloned ctsm
2) executed ./manage_externals/checkout_externals
3) changed in src/fates, then checked out a new branch from a remote
4) ran a test

My thoughts at the time, were that the testing system was calling manage_externals do check what it had in the configuration and what was actually checked out, and that there was some conflict with the new branch I checked-out inside fates

However, whatever problem I was having, was probably just me making a mistake.  I just tried those same steps and the test ran just fine.

My understanding now, is that if I use git to directly check-out new branches inside src/fates, then I do not have to make updates to the fates entry in Externals_CLM.cfg to run tests.

Sorry for the confusion from my side.

**Comment by ekluzek** [comment]
We talked about this briefly in a CSEG meeting, and decided to leave the current behavior in place. It is legit for manage_externals to die on a legitimate error. This does mean that users have to keep manage_externals functioning to at least give a valid status. So I think this means as part of the workflow to run "manage_externals" "--status" whenever you make changes that might screw it up so that your cases don't die when you create them.

**Comment by mnlevy1981** [comment]
Presuming `case.build` and `case.test` are the only user-facing script that calls `_save_build_provenance_cesm`, I'm in favor of the following:

1. `manage_externals` continues to return errors when it deems it appropriate
2. CESM does not set `BUILD_COMPLETE=TRUE` if `_save_build_provenance_cesm` fails
3. The build and test scripts have a `--skip-provenance-check` flag that skips this check (and maybe also let users set `SKIP_PROVENANCE_CHECK` in `.cime/config`?)

There are instances where we absolutely need to be sure that the provenance being reported is correct (otherwise why have the feature?), and for those cases we need `case.build` to prevent the user from continuing. There are other times when developers know their sandbox doesn't match what's in `Externals.cfg`, and they're just testing a new feature, so why would you even want the provence check to be run?

**Comment by jedwards4b** [comment]
For clarity the provenance check is not looking at Externals.cfg, it's querying repos to find out exactly what you do have.    Otherwise I agree with the approach @mnlevy1981 proposes.

**Comment by klindsay28** [comment]
comments from a user (me)...

Using a relatively recent codebase (cesm2_0_exp10e), I've cloned a case of
Cecile that uses a CESM sandbox of hers. When I build, I get an error from
checkout_externals. It appears that checkout_externals is attempting to
write to a mange_externals.log file that resides within her sandbox, and I
don't have write permission on that file.

I'm glad that the error is not truly fatal. Otherwise I would be unable to
easily clone her case.

So please consider the usecase of building a case when you don't have write
permission to $CIMEROOT.

FYI, the error message is:

ERROR: Command:
'/glade/p/work/hannay/cesm_tags/cesm2_0_exp10e/cime/../manage_externals/checkout_externals
--status --verbose' failed with error '[Errno 13] Permission denied:
u'/glade/p/work/hannay/cesm_tags/cesm2_0_exp10e/manage_externals.log'' from
dir '/glade/p/work/hannay/cesm_tags/cesm2_0_exp10e/cime/..'

**Comment by billsacks** [comment]
I like @mnlevy1981 's suggestions. The error printed if the provenance-gathering fails should suggest re-running with the `--skip-provenance-check` flag if you don't care about provenance information.

**Comment by billsacks** [comment]
@klindsay28 - the issue you raised was fixed in https://github.com/ESMCI/cime/pull/2444, which is in a cime version one past the one in Cecile's tag that you're using.

**Comment by klindsay28** [comment]
Thanks @billsacks.

**Comment by matt-long** [comment]
I have had an experience similar to @klindsay28's, though triggered by a different issue [manage-externals #100](https://github.com/ESMCI/manage_externals/issues/100) .  Perhaps if provenance-gathering fails, the user could just get a warning not an error.

**Comment by jedwards4b** [comment]
This is addressed in the PR #2543, an error is important, but that PR provides an override.
